### PR TITLE
[coreimage] Add new Image Dictionary Keys

### PR DESCRIPTION
**Pull request by VincentDondain**
- Fixes bug #59296: [coreimage] Some `kCI*`keys are not bound (https://bugzilla.xamarin.com/show_bug.cgi?id=59296)
- Generate a StrongDictionary for `CIImageInitializationOptions` to avoid manual code.
- Move `CGImageProperties Properties { get; set; }` to parent type `CIImageInitializationOptions` (avoid 2 strong dictionaries).
  Reason:
  Even though the headers give us an indication of which constructors should use some CIImage keys it's hard to apply that to all constructors consistently.
  We could have 1 strong dictionary per constructor (duplicate members) with just the exact members we know it supports (based on headers) however it's better to have a single strong dictionary and document the options because A might be available only in X today and Y next too next year.
- Fix `DictionaryContainer`'s `GetStrongDictionary` to return null and not throw if target StrongDictionary is not yet set.

Basically:

```
var options = new CIImageInitializationOptionsWithMetadata ();
Assert.That (options.Dictionary.Count, Is.EqualTo (0), "Count");
```

Would throw:

``` 
System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
 ----> System.ArgumentNullException : Value cannot be null.
```

- Add EditorBrowsableState.Advanced to CIImage's NSDictionary .ctor

### CODE REVIEW

File: src/CoreImage/CIImageInitializationOptions.cs [hunk 1]
```diff
@@ -34,53 +34,20 @@
...
-				return new CGImageProperties (dict);
-			}
-			set {
-				SetNativeValue (CIImage.CIImagePropertiesKey, value == null ? null : value.Dictionary, false);
-			}
-		}
-#endif
-	}
+	// Keeping 'CIImageInitializationOptionsWithMetadata' to avoid breaking change
+	public class CIImageInitializationOptionsWithMetadata : CIImageInitializationOptions {}
```
Comment (chamons) [inline src/CoreImage/CIImageInitializationOptions.cs]: This doesn't remove breaking change though. The properties property is gone?

File: src/CoreImage/CIImageInitializationOptions.cs [hunk 2]
```diff
@@ -34,53 +34,20 @@
 
 namespace XamCore.CoreImage {
 
-	public class CIImageInitializationOptions : DictionaryContainer
+	public partial class CIImageInitializationOptions
```
Comment (chamons) [inline src/CoreImage/CIImageInitializationOptions.cs]: Changing the base class of a public class is a breaking change?

Comment (VincentDondain) [reply src/CoreImage/CIImageInitializationOptions.cs]: Mmh it shouldn't be a breaking change I generate the strong dictionary here: https://github.com/xamarin/xamarin-macios/pull/3017/files#diff-14b0758943a42f6b156f42dc8a2f8116R1274

`CIImageInitializationOptions.g.cs` has `public partial class CIImageInitializationOptions : DictionaryContainer`

File: src/CoreImage/CIImageInitializationOptions.cs (hunk 1 above)
Comment (VincentDondain) [reply src/CoreImage/CIImageInitializationOptions.cs]: The property moved to the parent type: `CIImageInitializationOptions`.

See https://github.com/xamarin/xamarin-macios/pull/3017/files#diff-14b0758943a42f6b156f42dc8a2f8116R1279

Comment (VincentDondain) [reply src/CoreImage/CIImageInitializationOptions.cs]: `CIImageInitializationOptions.g.cs` has:

```
public CGImageProperties Properties {
	get {
		return GetStrongDictionary<CGImageProperties>(CIImageInitializationOptionsKeys.PropertiesKey);
	}
	set {
		SetNativeValue (CIImageInitializationOptionsKeys.PropertiesKey, value.Dictionary);
	}
}
```

Comment (VincentDondain) [reply src/CoreImage/CIImageInitializationOptions.cs]: And this fix: https://github.com/xamarin/xamarin-macios/pull/3017/files#diff-c0ecdbcb25b3bffa3adae45754621d51R240 is because the new generated code broke one of the monotouch tests.

Differs from https://github.com/xamarin/xamarin-macios/pull/3017/files#diff-55370b82cf373918db04c3b399b7cc56L76

File: src/CoreImage/CIImageInitializationOptions.cs [hunk 3]
```diff
@@ -34,53 +34,20 @@
...
 	}
 
-	public class CIImageInitializationOptionsWithMetadata : CIImageInitializationOptions
-	{
-#if !COREBUILD
-		public CIImageInitializationOptionsWithMetadata ()
-		{
-		}
-
-		public CIImageInitializationOptionsWithMetadata (NSDictionary dictionary)
```
Comment (rolfbjarne) [inline src/CoreImage/CIImageInitializationOptions.cs]: According to api-diff, this constructor is gone:

> public CIImageInitializationOptionsWithMetadata (Foundation.NSDictionary dictionary);

File: src/CoreImage/CIImageInitializationOptions.cs (hunk 2 above)
Comment (spouliot) [reply src/CoreImage/CIImageInitializationOptions.cs]: @chamons https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md <- it's in our wiki ;)

File: src/coreimage.cs [hunk 4]
```diff
@@ -941,6 +942,14 @@ interface CIFilterInputKey {
 		[Since (7,0)]
 		[Field ("kCIInputExtentKey", "+CoreImage")]
 		NSString Extent  { get; }
+
+		[iOS (11,0), TV (11,0), Mac (10,13)]
```
Comment (spouliot) [inline src/coreimage.cs]: @VincentDondain make are to run intro on mac/32 bits on 10.13 (in case they are only available on 64 bits).
The jenkins bots are on 10.12 and it's better verified before reaching wrench.

File: src/CoreImage/CIImageInitializationOptions.cs (hunk 2 above)
Comment (spouliot) [reply src/CoreImage/CIImageInitializationOptions.cs]: Jenkins bots also produce the API diff to verify

Comment (chamons) [reply src/CoreImage/CIImageInitializationOptions.cs]: @spouliot The ? was "this certainly looks like a break, what am I missing" not a question on if removing the base class is breaking. :)

File: src/CoreImage/CIImageInitializationOptions.cs (hunk 3 above)
Comment (VincentDondain) [reply src/CoreImage/CIImageInitializationOptions.cs]: Yea I had `CIImageInitializationOptionsWithMetadata` be generated at some point but moved away from that and forgot to re-add the .ctor.

API diff reports other breaking changes though:

screenshot 2017-11-17 11 15 01

However according to https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md this should be fine...

Comment (rolfbjarne) [reply src/CoreImage/CIImageInitializationOptions.cs]: api-diff will incorrectly report API that moved to a base class as a breaking change.

This is a known issue with the API diff.

**Comment by VincentDondain** [comment]
I ran the Mac [don't link](https://gist.github.com/VincentDondain/5352f20fdbbf1b4e1955cd68eacf0d9e) and [introspection-mac-unified-32](https://gist.github.com/VincentDondain/f19c0f1d7953040df1660f682f9f62ad) tests successfully on High Sierra.

*Note: I had to skip `NSTextInputContext` because of bug [#60762](https://bugzilla.xamarin.com/show_bug.cgi?id=60762).*

**Comment by VincentDondain** [comment]
Breaking changes are fixed, as mentioned in https://github.com/xamarin/xamarin-macios/pull/3017#discussion_r151754928 the rest are known API diff issues.
